smack-experimental/src/main/java/org/jivesoftware/smackx/sid/StableUniqueStanzaIdManager.java

```diff
@@ -0,0 +1,119 @@
+/**
+ *
+ * Copyright 2018 Paul Schaub
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jivesoftware.smackx.sid;
+
+import java.util.Map;
+import java.util.WeakHashMap;
+
+import org.jivesoftware.smack.ConnectionCreationListener;
+import org.jivesoftware.smack.Manager;
+import org.jivesoftware.smack.StanzaListener;
+import org.jivesoftware.smack.XMPPConnection;
+import org.jivesoftware.smack.XMPPConnectionRegistry;
+import org.jivesoftware.smack.filter.AndFilter;
+import org.jivesoftware.smack.filter.MessageTypeFilter;
+import org.jivesoftware.smack.filter.StanzaFilter;
+import org.jivesoftware.smack.filter.ToTypeFilter;
+import org.jivesoftware.smack.packet.Message;
+import org.jivesoftware.smack.packet.Stanza;
+import org.jivesoftware.smackx.disco.ServiceDiscoveryManager;
+import org.jivesoftware.smackx.sid.element.OriginIdElement;
+
+public final class StableUniqueStanzaIdManager extends Manager {
+
+    public static final String NAMESPACE = "urn:xmpp:sid:0";
+
+    private static final Map<XMPPConnection, StableUniqueStanzaIdManager> INSTANCES = new WeakHashMap<>();
+
+    // Filter for outgoing stanzas.
+    private static final StanzaFilter OUTGOING_FILTER = new AndFilter(
+            MessageTypeFilter.NORMAL_OR_CHAT_OR_HEADLINE,
+            ToTypeFilter.ENTITY_FULL_OR_BARE_JID);
+
+    // Listener for outgoing stanzas that adds origin-ids to outgoing stanzas.
+    private final StanzaListener stanzaListener = new StanzaListener() {
+        @Override
+        public void processStanza(Stanza packet) {
+            if (!(packet instanceof Message)) {
+                return;
+            }
+            Message message = (Message) packet;
+            if (!OriginIdElement.hasOriginId(message)) {
```

Comment:
Use (and maybe create) a WithOriginIdFilter, and use that filter in combination with the 'not' filter in the OUTGOING_FILTER, then the whole listener comes down to `OriginIdelement.addOriginId(message)`.